### PR TITLE
Trigonometry

### PR DESCRIPTION
**Pull request by DrunkenRandomWalker**
Add more basic functions to facilitate the modeling

1. only sin and cos, using Taylor expansion to approximate the true value.
2. add ZERO, TWO ... NINE to FPDecimal
3. factorial may have an overflow issue.
4. gamma function is not tested

### CODE REVIEW

File: packages/injective-math/src/fp_decimal/trigonometry.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
+use num::pow::Pow;
+
+use crate::fp_decimal::FPDecimal;
+
+impl FPDecimal {
+    pub fn _cos(mut x: FPDecimal) -> FPDecimal {
+        x = FPDecimal::_change_range(x);
+        FPDecimal::_sin(FPDecimal::PI / FPDecimal::TWO - x)
+        /*
+        let sign = x.sign;
```
Comment (gorgos) [inline packages/injective-math/src/fp_decimal/trigonometry.rs]: why this commented out code?

File: packages/injective-math/src/fp_decimal/trigonometry.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            - (x.pow(FPDecimal::SIX) / FPDecimal::SIX.factorial())
+            + (x.pow(FPDecimal::EIGHT) / FPDecimal::EIGHT.factorial())
+            - (x.pow(FPDecimal::TWO + FPDecimal::EIGHT) / (FPDecimal::TWO + FPDecimal::EIGHT).factorial())
+            + (x.pow(FPDecimal::FOUR + FPDecimal::EIGHT) / (FPDecimal::FOUR + FPDecimal::EIGHT).factorial());
+        output.sign = sign;
+        output
+        */
+    }
+
+    fn _taylro_expansion(x: FPDecimal) -> FPDecimal {
```
Comment (gorgos) [inline packages/injective-math/src/fp_decimal/trigonometry.rs]: wrong name? ->`_taylor_expansion`, also why stop at exactly 9 ?

File: packages/injective-math/src/fp_decimal/trigonometry.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        } else if FPDecimal::PI < x && x < pi_plus_pi_by_2 {
+            let mut output = FPDecimal::_taylro_expansion(x - FPDecimal::PI);
+            output.sign = 0;
+            output
+        } else {
+            let mut output = FPDecimal::_taylro_expansion(FPDecimal::PI * FPDecimal::TWO - x);
+            output.sign = 0;
+            output
+        }
+    }
```
Comment (gorgos) [inline packages/injective-math/src/fp_decimal/trigonometry.rs]:
```suggestion
    pub fn _sin(mut x: FPDecimal) -> FPDecimal {
        x = FPDecimal::_change_range(x);
        
        let pi = FPDecimal::PI;
        let pi_by_2 = pi / FPDecimal::TWO;
        let pi_plus_pi_by_2 = pi + pi_by_2;
        
        if x == FPDecimal::ZERO || x == pi {
            return FPDecimal::ZERO;
        }
        
        if pi_by_2 == x {
            return FPDecimal::ONE;
        }
        
        if pi_plus_pi_by_2 == x {
            return FPDecimal::ZERO - FPDecimal::ONE;
        }
        
        if FPDecimal::ZERO < x && x < pi_by_2 {
            return FPDecimal::_taylro_expansion(x);
        }
        
        if pi_by_2 < x && x < pi {
            return FPDecimal::_taylro_expansion(pi - x);
        }
        
        if FPDecimal::PI < x && x < pi_plus_pi_by_2 {
            let mut output = FPDecimal::_taylro_expansion(x - pi);
            output.sign = 0;
            return output;
        }
        
        let mut output = FPDecimal::_taylro_expansion(pi * FPDecimal::TWO - x);
        output.sign = 0;
        output
```

**Review by gorgos — CHANGES_REQUESTED** [review]
LGTM mostly 👍

File: packages/injective-math/src/fp_decimal/trigonometry.rs (hunk 1 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/trigonometry.rs]: These are old implementation, no longer needed

File: packages/injective-math/src/fp_decimal/trigonometry.rs (hunk 2 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/trigonometry.rs]: 1. 9 provides a good enough approximation to compute the sine(x), the relative error is less than 1%,
2. using a number greater than 9 may have an overflow issue.

File: packages/injective-math/src/fp_decimal/trigonometry.rs (hunk 1 above)
Comment (gorgos) [reply packages/injective-math/src/fp_decimal/trigonometry.rs]: so remove it instead of commenting out ?

**Review by gorgos — APPROVED** [review]
LGTM

**Pull request merged**